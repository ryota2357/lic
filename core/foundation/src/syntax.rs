pub mod ast;
pub mod token;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum LicoLanguage {}

impl rowan::Language for LicoLanguage {
    type Kind = SyntaxKind;
    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        SyntaxKind::from(raw)
    }
    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = rowan::SyntaxNode<LicoLanguage>;
pub type SyntaxToken = rowan::SyntaxToken<LicoLanguage>;
pub type SyntaxElement = rowan::SyntaxElement<LicoLanguage>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<LicoLanguage>;
pub type SyntaxElementChildren = rowan::SyntaxElementChildren<LicoLanguage>;
pub type PreorderWithTokens = rowan::api::PreorderWithTokens<LicoLanguage>;

macro_rules! syntax_kind {
    ($($variant:ident $(= [$($tt:tt)*])? $(@ $anchor:ident)?),* $(,)?) => {
        #[allow(bad_style)]
        #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
        #[repr(u8)]
        pub enum SyntaxKind {
            $(
                $variant
            ),*
        }

        #[macro_export]
        #[doc(hidden)]
        macro_rules! __token_kind_fast_accsess {
            $($(
                ($($tt)*) => { $crate::syntax::SyntaxKind::$variant };
            )?)*
        }
        pub use __token_kind_fast_accsess as T;

        impl From<SyntaxKind> for rowan::SyntaxKind {
            fn from(kind: SyntaxKind) -> Self {
                Self(kind as u16)
            }
        }

        impl From<rowan::SyntaxKind> for SyntaxKind {
            fn from(kind: rowan::SyntaxKind) -> Self {
                let n_variant = $crate::macros::count!($($variant)*) as u16;
                assert!(kind.0 <= n_variant, "bad SyntaxKind: {:?}", kind);
                // SAFETY: Ensured by the assert above.
                unsafe { ::core::mem::transmute(kind.0 as u8) }
            }
        }

        impl SyntaxKind {
            $($(const $anchor: Self = Self::$variant;)?)*
        }
    };
}
syntax_kind! {
    ERROR,

    // ↓ Leaf (token) ↓

    COMMENT,
    WHITESPACE,

    INT = [int] @START_LITERAL,
    FLOAT = [float],
    STRING = [string],
    TRUE = [true],
    FALSE = [false],
    NIL = [nil] @END_LITERAL,

    VAR_KW = [var] @START_KEYWORD,
    FUNC_KW = [func],
    IF_KW = [if],
    THEN_KW = [then],
    ELIF_KW = [elif],
    ELSE_KW = [else],
    FOR_KW = [for],
    WHILE_KW = [while],
    IN_KW = [in],
    DO_KW = [do],
    END_KW = [end],
    RETURN_KW = [return],
    BREAK_KW = [break],
    CONTINUE_KW = [continue],
    AND_KW = [and],
    OR_KW = [or],
    NOT_KW = [not] @END_KEYWORD,

    PLUS = [+] @START_PUNCT,
    MINUS = [-],
    STAR = [*],
    SLASH = [/],
    PERCENT = [%],
    AMP = [&],
    PIPE = [|],
    CARET = [^],
    TILDE = [~],
    BANG = [!],
    EQ = [=],
    LT = [<],
    GT = [>],
    DOT = [.],
    AT = [@],
    COMMA = [,],
    COLON = [:],
    OPENPAREN = ['('],
    CLOSEPAREN = [')'],
    OPENBRACE = ['{'],
    CLOSEBRACE = ['}'],
    OPENBRACKET = ['['],
    CLOSEBRACKET = [']'],
    ARROW = [->],
    BANGEQ = [!=],
    EQ2 = [==],
    LT2 = [<<],
    LTEQ = [<=],
    GT2 = [>>],
    GTEQ = [>=],
    DOT2 = [..] @END_PUNCT,

    IDENT = [ident],

    // ↓ Node (non-leaf) ↓

    PROGRAM,

    VAR_STMT,
    FUNC_STMT,
    FOR_STMT,
    WHILE_STMT,
    RETURN_STMT,
    BREAK_STMT,
    CONTINUE_STMT,
    EXPR_STMT,
    ATTR_STMT,

    IF_EXPR,
    DO_EXPR,
    CALL_EXPR,
    BINARY_EXPR,
    PREFIX_EXPR,
    INDEX_EXPR,
    FIELD_EXPR,
    METHOD_CALL_EXPR,
    PAREN_EXPR,

    LOCAL_VAR,
    LITERAL,
    ARRAY_CONST,
    TABLE_CONST,
    FUNC_CONST,

    ELSE_BRANCH,
    ELIF_BRANCH,

    PARAM_LIST,
    ARG_LIST,

    NAME,
    NAME_PATH,
    TABLE_FIELD,
    TABLE_FIELD_NAME_IDENT,
    TABLE_FIELD_NAME_EXPR,
}

const fn _static_assert_size() {
    const { assert!(core::mem::size_of::<SyntaxKind>() == 1) }
}

impl SyntaxKind {
    fn is_between(self, start: SyntaxKind, end: SyntaxKind) -> bool {
        (start as u8..=end as u8).contains(&(self as u8))
    }
    pub fn is_literal(self) -> bool {
        self.is_between(SyntaxKind::START_LITERAL, SyntaxKind::END_LITERAL)
    }
    pub fn is_keyword(self) -> bool {
        self.is_between(SyntaxKind::START_KEYWORD, SyntaxKind::END_KEYWORD)
    }
    pub fn is_punct(self) -> bool {
        self.is_between(SyntaxKind::START_PUNCT, SyntaxKind::END_PUNCT)
    }
    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::WHITESPACE | SyntaxKind::COMMENT)
    }
}
