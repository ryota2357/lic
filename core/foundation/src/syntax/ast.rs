use super::{LicoLanguage, SyntaxKind, SyntaxNode, SyntaxToken, T};
use rowan::ast::{support, AstChildren, AstNode};

macro_rules! __node_method {
    (child, $name:ident, $type:ty) => {
        pub fn $name(&self) -> Option<$type> {
            support::child(AstNode::syntax(self))
        }
    };
    (child, $name:ident, $type:ty, $idx:literal) => {
        pub fn $name(&self) -> Option<$type> {
            support::children(AstNode::syntax(self)).nth($idx)
        }
    };
    (children, $name:ident, $type:ty) => {
        pub fn $name(&self) -> AstChildren<$type> {
            support::children(AstNode::syntax(self))
        }
    };
    (token, $name:ident, $kind:tt) => {
        pub fn $name(&self) -> Option<SyntaxToken> {
            support::token(AstNode::syntax(self), T![$kind])
        }
    };
}
macro_rules! ast_node {
    (struct $name:ident for $kind:ident $({
        $( $method:ident: $method_kind:ident[$method_ret:tt $(, $method_ret_info:tt)?] ),* $(,)?
    })?) => {
        pub struct $name(SyntaxNode);
        impl AstNode for $name {
            type Language = LicoLanguage;
            fn can_cast(kind: <Self::Language as rowan::Language>::Kind) -> bool {
                kind == SyntaxKind::$kind
            }
            fn cast(node: rowan::SyntaxNode<Self::Language>) -> Option<Self> {
                if Self::can_cast(node.kind()) {
                    Some(Self(node))
                } else {
                    None
                }
            }
            fn syntax(&self) -> &rowan::SyntaxNode<Self::Language> {
                &self.0
            }
        }
        $(
            impl $name {
                $( __node_method! { $method_kind, $method, $method_ret $(, $method_ret_info)? })*
            }
        )?
    };
    (enum $name:ident for {
        $( $variant:ident($node:ty) ),* $(,)?
    }) => {
        pub enum $name {
            $($variant($node)),*
        }
        impl AstNode for $name {
            type Language = LicoLanguage;
            fn can_cast(kind: <Self::Language as rowan::Language>::Kind) -> bool {
                $(<$node as AstNode>::can_cast(kind))||*
            }
            fn cast(node: rowan::SyntaxNode<Self::Language>) -> Option<Self> {
                let kind = node.kind();
                $(
                    if <$node as AstNode>::can_cast(kind) {
                        let casted = <$node as AstNode>::cast(node).expect("Invalid `can_cast` implementation");
                        return Some(Self::$variant(casted));
                    }
                )*
                None
            }
            fn syntax(&self) -> &rowan::SyntaxNode<Self::Language> {
                match self {
                    $(Self::$variant(x) => x.syntax()),*
                }
            }
        }
    }
}

// Statement*
ast_node!(struct Program for PROGRAM {
    statements: children[Statement],
});

ast_node!(enum Statement for {
    Var(VarStmt),
    Func(FuncStmt),
    For(ForStmt),
    While(WhileStmt),
    Return(ReturnStmt),
    Break(BreakStmt),
    Continue(ContinueStmt),
    Expr(ExprStmt),
    Attr(AttrStmt),
});

// 'var' Name '=' Expr
ast_node!(struct VarStmt for VAR_STMT {
    var_token: token[var],
    name: child[Name],
    eq_token: token[=],
    initializer: child[Expression],
});

// 'func' NamePath ('->' Name)? ParamList
//   Program
// 'end'
ast_node!(struct FuncStmt for FUNC_STMT {
    func_token: token[func],
    name_path: child[NamePath],
    arrow_token: token[->],
    method_name: child[Name],
    param_list: child[ParamList],
    body: child[Program],
    end_token: token[end],
});

// 'for' Name 'in' Expr 'do'
//   Program
// 'end'
ast_node!(struct ForStmt for FOR_STMT {
    for_token: token[for],
    name: child[Name],
    in_token: token[in],
    iterable: child[Expression],
    do_token: token[do],
    body: child[Program],
    end_token: token[end],
});

// 'while' Expr 'do'
//   Program
// 'end'
ast_node!(struct WhileStmt for WHILE_STMT {
    while_token: token[while],
    condition: child[Expression],
    do_token: token[do],
    body: child[Program],
    end_token: token[end],
});

// 'return' Expr?
ast_node!(struct ReturnStmt for RETURN_STMT {
    return_token: token[return],
    expr: child[Expression],
});

// 'break'
ast_node!(struct BreakStmt for BREAK_STMT {
    break_token: token[break],
});

// 'continue'
ast_node!(struct ContinueStmt for CONTINUE_STMT {
    continue_token: token[continue],
});

// Expr
ast_node!(struct ExprStmt for EXPR_STMT {
    expr: child[Expression],
});

// '@' Name
ast_node!(struct AttrStmt for ATTR_STMT {
    at_token: token[@],
    name: child[Name],
});

ast_node!(enum Expression for {
    If(IfExpr),
    Do(DoExpr),
    Call(CallExpr),
    Binary(BinaryExpr),
    Prefix(PrefixExpr),
    Index(IndexExpr),
    Field(FieldExpr),
    MethodCall(MethodCallExpr),
    Paren(ParenExpr),
    LocalVar(LocalVar),
    Literal(Literal),
    ArrayConst(ArrayConst),
    TableConst(TableConst),
    FuncConst(FuncConst),
});

// 'if' Expr then
//   Program
//  ElifBranch*
//  ElseBranch?
// 'end'
ast_node!(struct IfExpr for IF_EXPR {
    if_token: token[if],
    condition: child[Expression],
    then_token: token[then],
    body: child[Program],
    elif_branches: children[ElifBranch],
    else_branch: child[ElseBranch],
    end_token: token[end],
});

// 'do'
//   Program
// 'end'
ast_node!(struct DoExpr for DO_EXPR {
    do_token: token[do],
    body: child[Program],
    end_token: token[end],
});

// Expr ArgList
ast_node!(struct CallExpr for CALL_EXPR {
    expr: child[Expression],
    arg_list: child[ArgList],
});

// Expr 'BinaryOp' Expr
ast_node!(struct BinaryExpr for BINARY_EXPR {
    lhs: child[Expression],
    rhs: child[Expression],
});
impl BinaryExpr {
    pub fn op(&self) -> Option<(SyntaxToken, BinaryOp)> {
        self.syntax()
            .children_with_tokens()
            .filter_map(|it| it.into_token())
            .find_map(|token| {
                #[rustfmt::skip]
                let op = match token.kind() {
                    T![+]   => BinaryOp::Add,
                    T![-]   => BinaryOp::Sub,
                    T![*]   => BinaryOp::Mul,
                    T![/]   => BinaryOp::Div,
                    T![%]   => BinaryOp::Mod,
                    T![<<]  => BinaryOp::Shl,
                    T![>>]  => BinaryOp::Shr,
                    T![..]  => BinaryOp::Concat,
                    T![==]  => BinaryOp::Eq,
                    T![!=]  => BinaryOp::Ne,
                    T![<]   => BinaryOp::Lt,
                    T![<=]  => BinaryOp::Le,
                    T![>]   => BinaryOp::Gt,
                    T![>=]  => BinaryOp::Ge,
                    T![and] => BinaryOp::And,
                    T![or]  => BinaryOp::Or,
                    T![&]   => BinaryOp::BitAnd,
                    T![|]   => BinaryOp::BitOr,
                    T![^]   => BinaryOp::BitXor,
                    T![=]   => BinaryOp::Assign,
                    _ => return None,
                };
                Some((token, op))
            })
    }
    pub fn op_token(&self) -> Option<SyntaxToken> {
        self.op().map(|(token, _)| token)
    }
    pub fn op_kind(&self) -> Option<BinaryOp> {
        self.op().map(|(_, kind)| kind)
    }
}
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Shl,
    Shr,
    Concat,
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
    And,
    Or,
    BitAnd,
    BitOr,
    BitXor,
    Assign,
}

ast_node!(struct PrefixExpr for PREFIX_EXPR {
    expr: child[Expression],
});
impl PrefixExpr {
    pub fn op(&self) -> Option<(SyntaxToken, PrefixOp)> {
        self.syntax()
            .children_with_tokens()
            .filter_map(|it| it.into_token())
            .find_map(|token| {
                #[rustfmt::skip]
                let op = match token.kind() {
                    T![+]   => PrefixOp::Plus,
                    T![-]   => PrefixOp::Minus,
                    T![not] => PrefixOp::Not,
                    T![~]   => PrefixOp::BitNot,
                    _ => return None,
                };
                Some((token, op))
            })
    }
    pub fn op_token(&self) -> Option<SyntaxToken> {
        self.op().map(|(token, _)| token)
    }
    pub fn op_kind(&self) -> Option<PrefixOp> {
        self.op().map(|(_, kind)| kind)
    }
}
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum PrefixOp {
    Plus,
    Minus,
    Not,
    BitNot,
}

// Expr '[' Expr ']'
ast_node!(struct IndexExpr for INDEX_EXPR {
    expr: child[Expression, 0],
    l_bracket_token: token['['],
    index: child[Expression, 1],
    r_bracket_token: token[']'],
});

// Expr '.' Name
ast_node!(struct FieldExpr for FIELD_EXPR {
    expr: child[Expression],
    dot_token: token[.],
    field: child[Name],
});

// Expr '->' Name ArgList
ast_node!(struct MethodCallExpr for METHOD_CALL_EXPR {
    expr: child[Expression],
    arrow_token: token[->],
    method_name: child[Name],
    arg_list: child[ArgList],
});

// '(' Expr ')'
ast_node!(struct ParenExpr for PAREN_EXPR {
    l_paren_token: token['('],
    expr: child[Expression],
    r_paren_token: token[')'],
});

// ident
ast_node!(struct LocalVar for LOCAL_VAR {
    ident_token: token[ident],
});

// int | float | string | 'true' | 'false' | nil
ast_node!(struct Literal for LITERAL);
impl Literal {
    pub fn token(&self) -> Option<SyntaxToken> {
        self.syntax()
            .children_with_tokens()
            .filter_map(|it| it.into_token())
            .find(|it| it.kind().is_literal())
    }
    pub fn kind(&self) -> Option<LiteralKind> {
        let token = self.token()?;
        #[rustfmt::skip]
        let kind = match token.kind() {
            T![int]    => LiteralKind::Int(token),
            T![float]  => LiteralKind::Float(token),
            T![string] => LiteralKind::String(token),
            T![true]   => LiteralKind::Bool(true),
            T![false]  => LiteralKind::Bool(false),
            T![nil]    => LiteralKind::Nil,
            _ => return None,
        };
        Some(kind)
    }
}
pub enum LiteralKind {
    Int(SyntaxToken),
    Float(SyntaxToken),
    String(SyntaxToken),
    Bool(bool),
    Nil,
}

// '[' ( Expr (',' Expr)* ','? )? ']'
ast_node!(struct ArrayConst for ARRAY_CONST {
    l_bracket_token: token['['],
    elements: children[Expression],
    r_bracket_token: token[']'],
});

// '[' TableField* ']'
ast_node!(struct TableConst for TABLE_CONST {
    l_brace_token: token['{'],
    fields: children[TableField],
    r_brace_token: token['}'],
});

// 'func' ParamList
//   Program
// 'end'
ast_node!(struct FuncConst for FUNC_CONST {
    func_token: token[func],
    param_list: child[ParamList],
    body: child[Program],
    end_token: token[end],
});

// 'else'
//   Program
ast_node!(struct ElseBranch for ELSE_BRANCH {
    else_token: token[else],
    body: child[Program],
});

// 'elif' Expr 'then'
//   Program
ast_node!(struct ElifBranch for ELIF_BRANCH {
    elif_token: token[elif],
    condition: child[Expression],
    then_token: token[then],
    body: child[Program],
});

// '(' Name* ')'
ast_node!(struct ParamList for PARAM_LIST {
    l_paren_token: token['('],
    params: children[Name],
    r_paren_token: token[')'],
});

// '(' Expr* ')'
ast_node!(struct ArgList for ARG_LIST {
    l_paren_token: token['('],
    args: children[Expression],
    r_paren_token: token[')'],
});

// ident
ast_node!(struct Name for NAME {
    ident_token: token[ident],
});

// Name (. NamePath)?
ast_node!(struct NamePath for NAME_PATH {
    name: child[Name],
    dot_token: token[.],
    child: child[NamePath],
});

// 'func'? TableFieldName = Expr
ast_node!(struct TableField for TABLE_FIELD {
    func_token: token[func],
    field_name: child[TableFieldName],
    eq_token: token[=],
    initializer: child[Expression],
});

// TableFieldNameIdent | TableFieldNameExpr
ast_node!(enum TableFieldName for {
    Ident(TableFieldNameIdent),
    Expr(TableFieldNameExpr),
});

// '[' Expr ']'
ast_node!(struct TableFieldNameExpr for TABLE_FIELD_NAME_EXPR {
    l_bracket_token: token['['],
    expr: child[Expression],
    r_bracket_token: token[']'],
});

// ident
ast_node!(struct TableFieldNameIdent for TABLE_FIELD_NAME_IDENT {
    ident_token: token[ident],
});
