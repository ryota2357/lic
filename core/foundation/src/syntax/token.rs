#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct Token {
    pub kind: TokenKind,
    pub len: u32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum TokenKind {
    /// "# comment"
    LineComment,

    /// Any whitespace character sequence.
    Whitespace,

    // Literal
    /// Integer literal. e.g. "42"
    Int { base: NumBase, empty_int: bool },
    /// Float literal. e.g. "3.14"
    Float { base: NumBase, empty_exponent: bool },
    /// String literal. e.g. "\"foo\""
    String {
        terminated: bool,
        quote_kind: QuoteKind,
    },
    /// "true"
    True,
    /// "false"
    False,
    /// "nil"
    Nil,

    /// Keyword (reserved identifiers)
    /// "var"
    Var,
    /// "func"
    Func,
    /// "if"
    If,
    /// "then"
    Then,
    /// "elif"
    Elif,
    /// "else"
    Else,
    /// "for"
    For,
    /// "while"
    While,
    /// "in"
    In,
    /// "do"
    Do,
    /// "end"
    End,
    /// "return"
    Return,
    /// "break"
    Break,
    /// "continue"
    Continue,
    /// "and"
    And,
    /// "or"
    Or,
    /// "not"
    Not,

    /// One character symbol.
    /// "+"
    Plus,
    /// "-"
    Minus,
    /// "*"
    Star,
    /// "/"
    Slash,
    /// "%"
    Percent,
    /// "&"
    Amp,
    /// "|"
    Pipe,
    /// "^"
    Caret,
    /// "~"
    Tilde,
    /// "!"
    Bang,
    /// "="
    Eq,
    /// "<"
    Lt,
    /// ">"
    Gt,
    /// "."
    Dot,
    /// "@"
    At,
    /// ","
    Comma,
    /// ":"
    Colon,
    /// "("
    OpenParen,
    /// ")"
    CloseParen,
    /// "{"
    OpenBrace,
    /// "}"
    CloseBrace,
    /// "["
    OpenBracket,
    /// "]"
    CloseBracket,

    // Two character symbol.
    /// "->"
    Arrow,
    /// "!="
    BangEq,
    /// "=="
    Eq2,
    /// "<<"
    Lt2,
    /// "<="
    LtEq,
    /// ">>"
    Gt2,
    /// ">="
    GtEq,
    /// ".."
    Dot2,

    /// Identifier that is not classified as a keyword or literal. e.g. "foo"
    Ident,

    /// Like the `Ident`, but containing invalid unicode codepoints.
    InvalidIdent,

    /// Unknown character, not expected by the lexer.
    Unknown,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum NumBase {
    /// Binary integer literal that starts with "0b".
    Binary = 2,
    /// Octal integer literal that starts with "0o".
    Octal = 8,
    /// Decimal integer literal.
    Decimal = 10,
    /// Hexadecimal integer literal that starts with "0x".
    Hexadecimal = 16,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum QuoteKind {
    /// Single quote string literal.
    Single,
    /// Double quote string literal.
    Double,
}
