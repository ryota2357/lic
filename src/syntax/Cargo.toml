[package]
name = "syntax"
version = "0.1.0"
edition = "2021"

[dependencies]
chumsky = { git = "https://github.com/zesterer/chumsky.git" }
