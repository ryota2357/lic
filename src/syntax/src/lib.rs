pub mod lex;
pub mod tree;

use chumsky::{input::Input, span::SimpleSpan, Parser};

type Span = SimpleSpan<usize>;

/// Parse a string into an Token list
pub fn parse_src_to_token(src: &str) -> Vec<(lex::Token<'_>, Span)> {
    lex::lexer().parse(src).unwrap()
}

/// Parse a Token list into a syntax tree
pub fn parse_token_to_syntax_tree<'tokens, 'src: 'tokens>(
    tokens: &'tokens [(lex::Token<'src>, Span)],
    eoi: std::ops::Range<usize>,
) -> (tree::node::Block<'src>, Span) {
    tree::parse::parser()
        .parse(tokens.spanned(eoi.into()))
        .unwrap()
}
