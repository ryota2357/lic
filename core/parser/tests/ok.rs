//! Generated by "../test_data/gen.ts". Do not edit by hand.
#![cfg_attr(rustfmt, rustfmt_skip)]
use foundation::syntax::SyntaxNode;
use insta::assert_snapshot;
macro_rules! test {
    ($name:ident, $path:literal) => {
        #[test]
        fn $name() {
            assert_snapshot!({
                let source = std::include_str!($path);
                let (green, err) = parser::parse(source, lexer::tokenize(source));
                assert!(err.is_empty(), "{:?}", err);
                format!("{:#?}", SyntaxNode::new_root(green))
            });
        }
    };
}
test!(empty_file, "../test_data/ok/000_empty_file.lico");
test!(comment, "../test_data/ok/001_comment.lico");
test!(var_stmt, "../test_data/ok/002_var_stmt.lico");
test!(func_stmt, "../test_data/ok/003_func_stmt.lico");
test!(for_stmt, "../test_data/ok/004_for_stmt.lico");
test!(while_stmt, "../test_data/ok/005_while_stmt.lico");
test!(if_stmt, "../test_data/ok/006_if_stmt.lico");
test!(do_stmt, "../test_data/ok/007_do_stmt.lico");
test!(return_stmt, "../test_data/ok/008_return_stmt.lico");
test!(break_stmt, "../test_data/ok/009_break_stmt.lico");
test!(assing_stmt, "../test_data/ok/010_assing_stmt.lico");
test!(call_stmt, "../test_data/ok/011_call_stmt.lico");
test!(method_call_stmt, "../test_data/ok/012_method_call_stmt.lico");
