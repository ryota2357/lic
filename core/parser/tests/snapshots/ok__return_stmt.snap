---
source: parser/tests/ok.rs
expression: "{\n    let source = std::include_str!(\"../test_data/ok/008_return_stmt.lico\");\n    let (green, err) = parser::parse(source, lexer::tokenize(source));\n    assert!(err.is_empty());\n    format!(\"{:#?}\", SyntaxNode :: new_root(green))\n}"
---
PROGRAM@0..23
  RETURN_STMT@0..7
    RETURN_KW@0..6 "return"
    WHITESPACE@6..7 "\n"
  RETURN_STMT@7..16
    RETURN_KW@7..13 "return"
    WHITESPACE@13..14 " "
    LITERAL@14..15
      INT@14..15 "1"
    WHITESPACE@15..16 "\n"
  RETURN_STMT@16..23
    RETURN_KW@16..22 "return"
    WHITESPACE@22..23 "\n"
